File: backend/src/main/java/com/triplevingt/game/engine/ClockEngine.java
package com.triplevingt.game.engine;

import java.util.List;
import tools.jackson.databind.JsonNode;

/**
 * Direct port of the original app's computeClockState (js/clock.js). Log
 * entry shape: {player: int, hit: boolean}. One shared sequential log like
 * Cricket — 3 darts per turn, current player derived from position.
 */
public final class ClockEngine {

    private ClockEngine() {}

    public static ClockState compute(List<String> players, JsonNode log) {
        int n = players.size();
        int[] hits = new int[n];
        int[] darts = new int[n];
        int len = log == null ? 0 : log.size();

        for (int i = 0; i < len; i++) {
            JsonNode t = log.get(i);
            int p = t.path("player").asInt();
            darts[p]++;
            if (t.path("hit").asBoolean(false)) hits[p]++;
        }

        int currentPlayer = n == 0 ? 0 : (len / 3) % n;
        int dartInTurn = len % 3;

        Integer winnerIndex = null;
        for (int p = 0; p < n; p++) { if (hits[p] >= 21) { winnerIndex = p; break; } }

        return new ClockState(hits, darts, currentPlayer, dartInTurn, winnerIndex, winnerIndex != null);
    }

    // The target for the `index`-th dart a player still needs (0-based).
    // Sequential mode (no clockOrder stored) is just index+1, 1-20 then 21
    // for the finish. Random mode looks up the shuffled draw stored on the
    // game at creation, so every player follows that same draw.
    public static int targetAt(List<Integer> clockOrder, int index) {
        if (clockOrder != null && !clockOrder.isEmpty()) return clockOrder.get(index);
        return index + 1;
    }
}

File: frontend/src/app/engine/clock-engine.ts
import { ClockLogEntry, Game } from '../core/api/models';

export interface ClockState {
  hits: number[];
  darts: number[];
  currentPlayer: number;
  dartInTurn: number;
  winnerIndex: number | null;
  finished: boolean;
}

// Direct port of computeClockState (js/clock.js) — one shared sequential
// log like Cricket, 3 darts per turn.
export function computeClockState(game: Game): ClockState {
  const n = game.players.length;
  const hits: number[] = game.players.map(() => 0);
  const darts: number[] = game.players.map(() => 0);
  const log = game.log as ClockLogEntry[];

  for (const t of log) {
    darts[t.player]++;
    if (t.hit) hits[t.player]++;
  }

  const currentPlayer = n === 0 ? 0 : Math.floor(log.length / 3) % n;
  const dartInTurn = log.length % 3;

  let winnerIndex: number | null = null;
  for (let p = 0; p < n; p++) { if (hits[p] >= 21) { winnerIndex = p; break; } }

  return { hits, darts, currentPlayer, dartInTurn, winnerIndex, finished: winnerIndex !== null };
}

export function finishLabel(mode: Game['finishMode']): string {
  if (mode === 'outer') return '25';
  if (mode === 'bull') return '50';
  return '25/50';
}

// The target for the `index`-th dart a player still needs (0-based).
export function clockTargetAt(game: Game, index: number): number {
  if (game.clockOrder && game.clockOrder.length) return game.clockOrder[index];
  return index + 1;
}

export function clockTargetLabel(game: Game, index: number): string {
  const num = clockTargetAt(game, index);
  if (num > 20) return finishLabel(game.finishMode);
  const prefix = game.clockMultiplier === 'double' ? 'D' : game.clockMultiplier === 'triple' ? 'T' : '';
  return prefix + num;
}
